### PR TITLE
Plugin logic separation

### PR DESCRIPTION
**Pull request by ch40s3k**

### CODE REVIEW

File: alamo_worker/plugins/__init__.py [hunk 1]
```diff
@@ -129,7 +150,8 @@ def dispatch(self, loop: 'ZMQEventLoop', payload: dict) -> Optional[dict]:
             plugin.obj for plugin in self.runner
             if _plugin_type in plugin.obj.supported_types()
         ]
-
+        ignored_fields = ['uuid', 'name', 'severity', 'enabled',
```
Comment (pnuckowski) [inline alamo_worker/plugins/__init__.py]: list to set

File: alamo_worker/plugins/__init__.py [hunk 2]
```diff
@@ -150,11 +172,23 @@ def dispatch(self, loop: 'ZMQEventLoop', payload: dict) -> Optional[dict]:
                 ) for plugin in coro_plugins
             ]
 
-            yield from asyncio.gather(*execute_tasks, loop=loop)
-            for plugin in no_coro_plugins:
-                yield from loop.run_in_executor(
-                    None, plugin.execute, payload
+            results = yield from asyncio.gather(*execute_tasks, loop=loop)
```
Comment (pnuckowski) [inline alamo_worker/plugins/__init__.py]: Maybe:
```
execute_tasks.append(
    loop.run_in_executor(None, plugin.execute, payload)
)
results = yield from asyncio.gather(*execute_tasks, loop=loop)
# so on ?
```

Comment (ch40s3k) [inline alamo_worker/plugins/__init__.py]: good 👍

File: alamo_worker/plugins/druid.py [hunk 3]
```diff
@@ -31,24 +31,116 @@
...
+    def set_check_status(self):
+        return dict(result=dict(
+            status=self.status, message=', '.join(self.failed_messages),
+            hysteresis=self.hysteresis)
+        )
+
+    @stats.increment(metric_name='druid.validate')
+    @stats.timer(metric_name='druid.validate')
+    def validate(self, trigger):
+        if self.status != RESULT_OK:
```
Comment (pnuckowski) [inline alamo_worker/plugins/druid.py]: do manager need to call `validate` if result is already failed/unknown ? maybe manager should handle this.

File: alamo_worker/plugins/druid.py [hunk 4]
```diff
@@ -31,24 +31,116 @@
...
+            self.failed_messages.append(self.message)
+        else:
```
Comment (pnuckowski) [inline alamo_worker/plugins/druid.py]: this method is quite complex. maybe drop `else` and add return inside first if.

File: alamo_worker/plugins/druid.py [hunk 5]
```diff
@@ -31,24 +31,116 @@
...
+                lambda x: condition(x, threshold), datapoints
+            ))
+
+            hystfail_values = list(filter(
+                lambda x: condition(x, hyst_threshold), datapoints
+            ))
+            if failed_values:
+                self.status = self.hysteresis = RESULT_FAILED
+                values = ', '.join(
+                    [str(x) for x in failed_values[-int(debounce):]]
```
Comment (pnuckowski) [inline alamo_worker/plugins/druid.py]: this should be inside graphite plugin.
Result should be initialized with proper data ( any None's, dropped last value should be in plugin)

File: alamo_worker/plugins/__init__.py [hunk 6]
```diff
@@ -112,10 +123,20 @@ def prepare_meta(self, check):
         for trigger in check['triggers']:
             trigger['meta'] = dict(
                 links=dict(
-                    trigger_url=dict(type='link', href=trigger['url'])
+                    trigger_url=dict(type='link', href=trigger.get('url', ''))
                 )
             )
 
+    def update(self, to_update: Dict, update_from: Dict) -> Dict:
+        """Workaround to update nested dicts."""
```
Comment (pnuckowski) [inline alamo_worker/plugins/__init__.py]: what is going on here?
 do we need this? maybe another solution or approach?

File: alamo_worker/plugins/druid.py [hunk 7]
```diff
@@ -31,24 +31,116 @@
...
+
+    def __init__(self, plugin_name, data, message='', status=RESULT_OK):
+        self.plugin_name = plugin_name
+        self.data = data
+        self.message = message
+        self.status = status
+        self.hysteresis = RESULT_OK
+        self.failed_messages = []
+
+    def set_check_status(self):
```
Comment (pnuckowski) [inline alamo_worker/plugins/druid.py]: the name of this is misleading. Is does not "set" status it returns or ever build trigger status.

File: alamo_worker/plugins/druid.py [hunk 8]
```diff
@@ -31,24 +31,116 @@
...
+                message = INVALID_THRESHOLD.format(
+                    trigger.get('threshold', '<unknown>'),
+                    trigger['id'], e
+                )
+                logger.error(message)
+                self.status = self.hysteresis = RESULT_UNKNOWN
+                self.failed_messages.append(message)
+                return self.set_check_status()
+            except KeyError:
+                message = INVALID_OPERATOR.format(
```
Comment (pnuckowski) [inline alamo_worker/plugins/druid.py]: I'm wondering if could create some 'API' to properly stop validation process and set result and return it.

example
missing or intentionally `self.failed_messages.append(message)` is missing. So i don't know is it ok or not?

File: alamo_worker/plugins/graphite.py [hunk 9]
```diff
@@ -48,19 +43,38 @@ class GraphitePlugin(BasePlugin):
...
-        return settings.GRAPHITE_TYPES
+    def prepare_values(self, env, frequency):
+        if not self.data:
+            stats.increment.incr('graphite.fail.{}.no_results'.format(env))
 
-    def init(self, *args):
-        super().init(*args)
+        for result in self.data:
+            datapoints = result['datapoints']
+            values = [float(t[0])
```
Comment (pnuckowski) [inline alamo_worker/plugins/graphite.py]: Those values should be already  filtered in plugin

File: alamo_worker/plugins/graphite.py [hunk 10]
```diff
@@ -126,90 +99,70 @@ def add_meta(self, check, trigger, env, threshold=None):
...
+    @stats.increment(metric_name='graphite.validate')
+    @stats.timer(metric_name='graphite.validate')
+    def validate(self, trigger):
+        env = trigger.get('environment', 'test')
+        frequency = int(trigger['frequency'])
+        expected_num_hosts = trigger.get(
+            'expected_num_hosts', 0)
+        threshold = 0
+
+        if self.status != RESULT_OK:
```
Comment (pnuckowski) [inline alamo_worker/plugins/graphite.py]: if manager knows that result is already failed/unknown do we need to call validate.

or 
drop else and add return inside if

File: alamo_worker/plugins/graphite.py [hunk 11]
```diff
@@ -126,90 +99,70 @@ def add_meta(self, check, trigger, env, threshold=None):
...
+                return self.set_check_status()
             except KeyError:
                 message = INVALID_OPERATOR.format(
                     trigger.get('condition'),
                     list(self.condition_map.keys())
                 )
                 logger.error(message)
-                self.set_status(
-                    trigger, RESULT_UNKNOWN, message, hysteresis=RESULT_UNKNOWN
+                self.hysteresis = self.status = RESULT_UNKNOWN
```
Comment (pnuckowski) [inline alamo_worker/plugins/graphite.py]: Already mentioned, some api for messaging?

File: alamo_worker/plugins/http_check.py [hunk 12]
```diff
@@ -47,6 +47,84 @@ class NoSuchService(Exception):
...
+            self.validate_status(trigger['status_code'])
+            contains = str(trigger['contains'])
+            contains = bytes(contains, 'utf-8').decode('unicode_escape')
+            if contains not in self:
+                self.status = RESULT_FAILED
+                self.failed_messages.append(
+                    CONTAINS_TEMPLATE.format(
+                        contains, self.url)
+                )
+            if (parsed_url.scheme == 'service'
```
Comment (pnuckowski) [inline alamo_worker/plugins/http_check.py]: it look like stairs to heaven :D `if -> and -> and`

File: alamo_worker/plugins/tests/test_graphite_plugin.py [hunk 13]
```diff
@@ -74,6 +74,21 @@ def setUpClass(cls):
     def setUp(self):
         self.loop = asyncio.get_event_loop()
         self.graphite = GraphitePlugin()
+        self.graphite_result = GraphiteResult(
+            plugin_name='GraphitePlugin', data={})
+        self.ignored_fields = [
+            'uuid', 'name', 'severity', 'enabled', 'url', 'tags']
+
+    def prepare_fields(self, check):
```
Comment (pnuckowski) [inline alamo_worker/plugins/tests/test_graphite_plugin.py]: Duplicated method?:)

File: alamo_worker/plugins/graphite.py [hunk 14]
```diff
@@ -218,41 +171,124 @@ def process(self, check, data, env):
...
+                    url,
+                    params={
+                        'target': metric,
+                        'from': settings.GRAPHITE_TIME_RANGE,
+                        'format': 'json'
+                    }
+                )
+
+                response.raise_for_status()
+                data = yield from response.json()
```
Comment (pnuckowski) [inline alamo_worker/plugins/graphite.py]: Maybe we should merge all those timeseries into one and group it into `{'timestamp': ['list of values']}`
This code was already written and used in check-tester.

File: alamo_worker/plugins/__init__.py (hunk 2 above)
Comment (ch40s3k) [inline alamo_worker/plugins/__init__.py]: raise TypeError("coroutines cannot be used with run_in_executor()")

File: alamo_worker/plugins/druid.py [hunk 15]
```diff
@@ -118,94 +210,17 @@ def execute(self, check):
                 params['filter'] = filters
 
             ts = query(**params)
-            self.process(check, ts.result)
-
+            return DruidResult(
```
Comment (pnuckowski) [inline alamo_worker/plugins/druid.py]: Can we prepare druid result in the way that the same login used in graphite could analyze druid results.

File: alamo_worker/plugins/druid.py (hunk 4 above)
Comment (ch40s3k) [inline alamo_worker/plugins/druid.py]: It won't be needed if we gonna validate in manager if status is not RESULT_OK

File: alamo_worker/plugins/__init__.py [hunk 16]
```diff
@@ -1,17 +1,32 @@
 # -*- coding: utf-8 -*-
 import asyncio
 import logging
+import operator
+from collections import OrderedDict
+
+from typing import List
```
Comment (pnuckowski) [inline alamo_worker/plugins/__init__.py]: Here typing... and in line 14 also typing :)

File: alamo_worker/plugins/druid.py [hunk 17]
```diff
@@ -31,25 +35,78 @@
...
+
+    @stats.increment(metric_name='druid.validate')
+    @stats.timer(metric_name='druid.validate')
+    def validate(self, trigger):
+        try:
+            threshold = float(trigger.get('threshold'))
+            condition = self.condition_map[trigger.get('condition')]
+            hysteresis_value = float(trigger.get('hysteresis', "0.0"))
+
+        except Exception as e:
```
Comment (pnuckowski) [inline alamo_worker/plugins/druid.py]: Why broad `Exception`? Maybe `ValueError` and `TypeError`?

File: alamo_worker/plugins/druid.py [hunk 18]
```diff
@@ -79,6 +136,13 @@ def build_filter(filters: list):
 
         return dim
 
+    def collect_result(self, result: List[Dict]):
+        datapoints = [
+            list(k['result'].values())[0] for k in result[-2:-1]
```
Comment (pnuckowski) [inline alamo_worker/plugins/druid.py]: Add please some comment why we have `result[-2:-1]`

Comment (ch40s3k) [inline alamo_worker/plugins/druid.py]: Przedostatni element, w sumie wystarczyłoby [-2]

File: alamo_worker/plugins/http_check.py [hunk 19]
```diff
@@ -47,6 +47,82 @@ class NoSuchService(Exception):
...
 
+class HttpCheckResult(BaseResult):
+    __slots__ = ('plugin_name', 'status_code', 'text', 'num_instances',
+                 'message', 'status', 'failed_messages', 'url')
+
+    def __init__(self, plugin_name, status_code, text, num_instances,
+                 message='', status=RESULT_OK, url=''):
+        self.plugin_name = plugin_name
+        self.status_code = status_code
+        self.text = text
```
Comment (pnuckowski) [inline alamo_worker/plugins/http_check.py]: `content` :)

File: alamo_worker/plugins/http_check.py [hunk 20]
```diff
@@ -47,6 +47,82 @@ class NoSuchService(Exception):
...
+                regexp_fail = not bool(pattern.search(self.text))
+            except re.error:
+                regexp_fail = True
+            if regexp_fail:
+                return False
+            return True
+        return True
+
+    def healthy_instances(self, instances):
+        if self.num_instances < instances:
```
Comment (pnuckowski) [inline alamo_worker/plugins/http_check.py]: `return self.num_instances < instances`?

File: alamo_worker/plugins/druid.py (hunk 17 above)
Comment (ch40s3k) [inline alamo_worker/plugins/druid.py]: see handle_exception function

File: alamo_worker/plugins/__init__.py [hunk 21]
```diff
@@ -1,17 +1,30 @@
 # -*- coding: utf-8 -*-
 import asyncio
 import logging
-import pkg_resources
+import operator
+from abc import ABCMeta
+from collections import OrderedDict
+from copy import deepcopy
```
Comment (deejay1) [inline alamo_worker/plugins/__init__.py]: wrong import order? ;)

File: alamo_worker/plugins/__init__.py [hunk 22]
```diff
@@ -1,17 +1,30 @@
...
 from datetime import datetime
-from stevedore import NamedExtensionManager
-from typing import Optional
 
+import pkg_resources
 from pytz import utc
+from stevedore import NamedExtensionManager
+from typing import List, Optional, Dict
```
Comment (deejay1) [inline alamo_worker/plugins/__init__.py]: typing is a core lib, move it to the other group

File: alamo_worker/plugins/__init__.py [hunk 23]
```diff
@@ -29,6 +42,26 @@
 }
 
 
+class BaseResult(metaclass=ABCMeta):
+    """Base Result object."""
+
+    def validate(self, trigger) -> Dict:
```
Comment (deejay1) [inline alamo_worker/plugins/__init__.py]: missign @abstractmethod decorators?

Comment (deejay1) [inline alamo_worker/plugins/__init__.py]: also missing type info for parameter

File: alamo_worker/plugins/druid.py [hunk 24]
```diff
@@ -79,6 +136,16 @@ def build_filter(filters: list):
 
         return dim
 
+    def collect_result(self, result: List[Dict]):
+        """Takes only second to last element. Because of druid specification,
```
Comment (deejay1) [inline alamo_worker/plugins/druid.py]: multiline doc strings should start on a new line

File: alamo_worker/plugins/http_check.py [hunk 25]
```diff
@@ -47,6 +47,80 @@ class NoSuchService(Exception):
...
+            regexp_fail = False
+            try:  # if full-text match failed, fall back to regexp
+                pattern = re.compile(key)
+                regexp_fail = not bool(pattern.search(self.content))
+            except re.error:
+                regexp_fail = True
+            if regexp_fail:
+                return False
+            return True
+        return True
```
Comment (deejay1) [inline alamo_worker/plugins/http_check.py]: so many return values.....

File: alamo_worker/plugins/__init__.py (hunk 21 above)
Comment (ch40s3k) [inline alamo_worker/plugins/__init__.py]: It's good :) first imports then froms

File: alamo_worker/plugins/druid.py (hunk 24 above)
Comment (ch40s3k) [inline alamo_worker/plugins/druid.py]: ./alamo_worker/plugins/druid.py:143:1: H404  multi line docstring should start without a leading new line

File: alamo_worker/plugins/http_check.py (hunk 25 above)
Comment (pnuckowski) [inline alamo_worker/plugins/http_check.py]:
```
        contains = key in self.content
        if not contains:
            try:  # if full-text match failed, fall back to regexp
                pattern = re.compile(key)
                contains = bool(pattern.search(self.content))
            except re.error:
                contains = False
        return contains
```
@deejay1 @h00dy @benethor  ?

Comment (ch40s3k) [inline alamo_worker/plugins/http_check.py]: rewritten

**Pull request merged**